File: questions/235.LowestCommonAncestorOfABinarySearchTree.js
function TreeNode(val) {
  this.val = val
  this.left = null
  this.right = null
}

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
const lowestCommonAncestor = function (root, p, q) {
  let current = root

  while (current) {
    if (p.val > current.val && q.val > current.val) {
      current = current.right
    } else if (p.val < current.val && q.val < current.val) {
      current = current.left
    } else {
      return current
    }
  }
}
const tree = new TreeNode(6)
tree.left = new TreeNode(2)
tree.left.left = new TreeNode(0)
tree.left.right = new TreeNode(4)
tree.left.right.left = new TreeNode(3)
tree.left.right.right = new TreeNode(5)

tree.right = new TreeNode(8)
tree.right.left = new TreeNode(7)
tree.right.right = new TreeNode(9)

console.log(lowestCommonAncestor(tree, tree.left, tree.right).val)
console.log(lowestCommonAncestor(tree, tree.left, tree.left.right).val)

File: questions/235.LowestCommonAncestorOfABinarySearchTree.go
package questions

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	current := root

	for current != nil {
		if p.Val > current.Val && q.Val > current.Val {
			current = current.Right
		} else if p.Val < current.Val && q.Val < current.Val {
			current = current.Left
		} else {
			return current
		}
	}

	return root
}

//tree := &TreeNode{Val: 6}
//tree.Left = &TreeNode{Val: 2}
//tree.Left.Left = &TreeNode{Val: 0}
//tree.Left.Right = &TreeNode{Val: 4}
//tree.Left.Right.Left = &TreeNode{Val: 3}
//tree.Left.Right.Right = &TreeNode{Val: 5}
//
//tree.Right = &TreeNode{Val: 8}
//tree.Right.Left = &TreeNode{Val: 7}
//tree.Right.Right = &TreeNode{Val: 9}
//
//
//result := lowestCommonAncestor(tree, tree.Left, tree.Right)
//
//fmt.Printf("%+v", result)
